packages/lending/src/adapter.ts

```diff
@@ -185,6 +191,335 @@ export class Adapter extends common.Web3Toolkit {
+   * @returns {Promise<OperationOutput>} The result including the destination amount,
+   * after portfolio, potential errors, and logic operations.
+   *
+   * 1. validate user inputs
+   * 2. zap-supply zap token
+   * 3. leverage by collateral token
+   */
+  async openByCollateral({
+    account,
+    portfolio,
+    zapToken,
+    zapAmount,
+    collateralToken,
+    collateralAmount,
+    debtToken,
+    slippage = defaultSlippage,
+  }: OpenOperationInput): Promise<OperationOutput> {
+    const output: OperationOutput = {
+      destAmount: '0',
+      afterPortfolio: portfolio.clone(),
+      logics: [],
+    };
+    try {
+      // 1. ---------- validate params ----------
+      const srcCollateral = portfolio.findSupply(collateralToken);
+      if (!srcCollateral) throw new OperationError('collateralToken', 'UNSUPPORTED_TOKEN');
+
+      const destBorrow = portfolio.findBorrow(debtToken);
+      if (!destBorrow) throw new OperationError('debtToken', 'UNSUPPORTED_TOKEN');
+
+      const supplyAmountDelta = new common.TokenAmount(collateralToken.wrapped, collateralAmount).sub(
+        srcCollateral.balance
+      );
+
+      // 1-1. ---------- zap swap ----------
+      const supplyInput = new common.TokenAmount(collateralToken.wrapped, '0');
+      if (Number(zapAmount) > 0) {
+        if (zapToken.is(collateralToken.wrapped)) {
+          // 2-1. exclude native zapToken
+          supplyInput.set(zapAmount);
+        } else {
+          const swapper = this.findSwapper([zapToken, collateralToken]);
+          let swapQuotation: SwapperQuoteFields;
+          try {
+            swapQuotation = await swapper.quote({
+              input: new common.TokenAmount(zapToken, zapAmount),
+              tokenOut: collateralToken.wrapped,
+              slippage,
+            });
+          } catch {
+            throw new OperationError('zapAmount', 'NO_ROUTE_FOUND_OR_PRICE_IMPACT_TOO_HIGH');
+          }
+          const swapTokenLogic = swapper.newSwapTokenLogic(swapQuotation);
+          output.logics.push(swapTokenLogic);
+          // 1-1-1. the supply amount is the swap quotation output
+          supplyInput.set(swapQuotation.output);
+        }
+        output.afterPortfolio.supply(supplyInput.token, supplyInput.amount);
+      }
+
+      // 1-2. collateralAmount must greater than init amount + zap-supply amount
+      if (Number(collateralAmount) <= Number(srcCollateral.balance) + Number(supplyInput.amount)) {
+        throw new OperationError('collateralAmount', 'COLLATERAL_AMOUNT_EXCEEDED');
+      }
+
+      // 2. leverage by collateral
+      const leverageAmount = supplyAmountDelta.sub(supplyInput).amount;
+      const leverageOutput = await this.leverageByCollateral({
+        account,
+        portfolio: output.afterPortfolio,
+        srcToken: collateralToken.wrapped,
+        srcAmount: leverageAmount,
+        destToken: debtToken,
+      });
```

Comment:
since `leverageXXX` does not handle native token and we can not let `zapWithdraw` handle native token (cause withdraw is followed by flashloan repay), open & close need to handle native token. But `debtToken` needs to be changed to `debtToken.wrapped` here.